### PR TITLE
Move the execution of the retention purge_jobs to the main worker

### LINKED ISSUE #9927: Retention policy purge jobs fail if `run_background_tasks_on` is set to something other than master
If `run_background_tasks_on` is set to anything other than `master`, Synapse will try to run background jobs on a worker, including the message retention policies support's purge jobs. However, these jobs end up calling the `purge_history` function of the storage layer (in the `PurgeEventsStorage` class) which doesn't exist on a worker, therefore the jobs fail with `AttributeError: 'GenericWorkerSlavedStore' object has no attribute 'purge_history'`.

This bug was introduced in https://github.com/matrix-org/synapse/pull/8513

cc @iansinclair

### LINKED ISSUE #13604: MSC1763 retention purge_jobs fail to run when run_background_tasks_on is configured to be a synapse.app.generic_worker
### Description

m.room.retention state events are correctly filtering out events in the endpoints, but the purge jobs aren't correctly running and the events are still in the database.

Seeing this error log which points pretty clearly at the issue
`{"log":"[purge] failed","namespace":"synapse.handlers.pagination","level":"ERROR","time":1661238458.09,"request":"_purge_history-0","server_name":"beeper.com","exc_type":"AttributeError","exc_value":"'GenericWorkerSlavedStore' object has no attribute 'purge_history'"}`

Can we just add `PurgeEventsStore` to `GenericWorkerSlavedStore` or can only the main process purge events? 🤔 

### Steps to reproduce

Relevant config:
```
run_background_tasks_on: synapse-background-worker-0

retention:
  enabled: true
```
so it's just the default daily job.

`synapse-background-worker-0` is the hostname of a `synapse.app.generic_worker` process configured to run background jobs.

### Homeserver

beeper.com

### Synapse Version

1.64.0 (Beeper fork)

### Installation Method

_No response_

### Platform

Running in containers on Kubernetes.

### Relevant log output

```shell
{"log":"[purge] failed","namespace":"synapse.handlers.pagination","level":"ERROR","time":1661238458.33,"request":"_purge_history-7","server_name":"beeper.com","exc_type":"AttributeError","exc_value":"'GenericWorkerSlavedStore' object has no attribute 'purge_history'"}
{"log":"[purge] Attempting to purge messages in room !XPxwZVXpPnNWmrJQwB:beeper.com","namespace":"synapse.handlers.pagination","level":"INFO","time":1661238458.31,"request":"purge_history_for_rooms_in_range-0","server_name":"beeper.com"}
```
```


### Anything else that would be useful to know?

_No response_

### PR DESCRIPTION
**Pull request by bradtgmurray**
Fixes #9927 

Signed-off-by: Brad Murray [brad@beeper.com](mailto:brad@beeper.com)

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). 
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct

### CODE REVIEW

File: changelog.d/13632.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix the running of MSC1763 retention purge_jobs in multi-worker setups by only running them on the main worker. Contributed by Brad @ Beeper.
```
Comment (clokep) [inline changelog.d/13632.bugfix:1]: I think this is really "when you have a background worker configured"?

Comment (bradtgmurray) [inline changelog.d/13632.bugfix:1]:
```suggestion
Fix the running of MSC1763 retention purge_jobs in deployments with background jobs running on a worker by forcing them back onto the main worker. Contributed by Brad @ Beeper.
```

**Review by clokep — APPROVED** [review]
👍  Looks good, thanks.

**Comment by verymilan** [comment]
ah, so this is why my replication times out due to the main process being overloaded.

**Pull request merged**